File: moon-breaker/moonbreaker.cpp
#include <iostream>
#include <cmath>
using namespace std;

int TtoB(string text) {
    long int result = 0;
    for (long int i = 0; i < text.length(); i++) {
        result += (long int)text[i];
    }
    return result;
}

// convert binary to int
int BtoI(string binary) {
    long int result = 0;
    for (long int i = 0; i < binary.length(); i++) {
        result += (long int)binary[i] * pow(2, binary.length() - i - 1);
    }
    return result;
}

// number of caracters in a int
int nb_caracters(long long int n) {
    int result = 0;
    while (n > 0) {
        n /= 10;
        result++;
    }
    return result;
}

long long int MakeKey (string bina, int key)
{
    long long int basse = 1;
    long int sup = BtoI(bina);
    for (long int x = 2 ; x < sizeof(bina)*key + 20 ; x = x  + 1)
    {
        basse = basse * int(bina[x%(sizeof(bina))])+1;
    }


    while (true)
    {
        if (sup > basse)
        {
            sup = sup / 2;
        }
        else if (sup*10 < basse)
        {
            sup = sup * 2;
        }
        else
        {
            break;
        }
    }
    

    basse = basse - sup;

    
    
    while (true)
    {
        cout << "while " << basse << endl;
        
        if (nb_caracters(basse) > 10)
        {
            basse = basse / 10;
        }
        else if (nb_caracters(basse) < 10)
        {
            basse = basse * 10;
        }
        else
        {
            break;
        }
    }

    return basse;
}

long long int moonbreaker(string text, int key)
{
    return MakeKey(TtoB(text) + "1", key);
}

int main()
{
    string text;
    cout << "Texte: ";
    cin >> text;
    cin.ignore();
    long long int sortie = moonbreaker(text, 2);
    cout << sortie << endl;
    return 0;
}

File: moon-breaker/moonbreaker.py
'''    _             _
  ___ | | _   _   __| | _ __   ___
 / _ \| || | | | / _` || '__| / _ |
|  __/| || |_| || (_| || |   |  __/
 \___||_| \__, | \__,_||_|    \___|
          |___/
___________________________________

 - codé en : UTF-8
 - langage : python3
 - GitHub  : github.com/elydre
 - Licence : GNU GPL v3
'''

version = "0.5"

BtoI = lambda bina: int(bina, 2)
TtoB = lambda text: "".join(format(ord(i), "08b") for i in text)

def MakeKey(bina,key):
    basse = 1
    sup = BtoI(bina)
    for x in range(2, len(bina) * key + 20):
        basse *= int(bina[x % (len(bina))]) + 1

    while True:
        if sup > basse: sup /= 2
        elif sup * 10 < basse: sup *= 2
        else: break
    
    basse -= sup
    
    while True:
        if len(str(basse)) > 10: basse //= 10
        elif len(str(basse)) < 10: basse *= 10
        else: break
    
    return int(basse)


def moonbreaker(txt, key = 2):
    return MakeKey(f"{TtoB(txt)}1", key)

if __name__ == "__main__":
    print(moonbreaker(input("Texte (str) : "), int(input("Key   (int) : "))))